### ISSUE TITLE
Can't play recordings

### ISSUE THREAD

**Issue by jabbors** [rsiebert/TVHClient#276]
I think the client tries to open an incorrect URL to a recording. Opening any recording with either VLC or BSPlayer fails. VLC outputs the URL to the recoding if it fails to open it and the below screenshot is what I got.

![screenshot_2016-11-20-08-54-02](https://cloud.githubusercontent.com/assets/1805094/20642559/7938bc4e-b41a-11e6-9a22-ce7b34596b4a.png)

With [TvhClient](https://itunes.apple.com/us/app/tvhclient/id638900112?mt=8&ign-mpt=uo%3D4) on iOS I can play the recoding just fine and the URL to the recording I obtained from that client is: `http://*************@*************:9981/tvh/dvrfile/175c6fe5f3cb4d552c9670d0b3616a34`

Looking at the dvr files on the tvheadend server I found a file with name "175c6fe5f3cb4d552c9670d0b3616a34" but not any file with the "84BBE..." fingerprint.

**Comment by Philipp91** [comment]
I also can't play recordings, with no error message though: I click a recording and TVHClient opens its internal player, which displays a circular loading indicator. After a few seconds, that one disappears and the sceen is just black with the info icon at the bottom and the back and menu icons at the top.  There is no error message.
When I click info, it displays a correct link. Typing this link into VLC on a PC works fine. Unfortunately the link can't be copied to clipboard on the device.

Also, how can I open it with VLC? I installed "VLC for Android", but TVHClient still opens only its internal player.

**Comment by rsiebert** [comment]
Interesting. I did not change the player logic. The URL with the ticket id is working fine here. 

I check if I could add an option to use an alternative url like the one shown above. 

The tvhclient app has no internal player. It passes the URL to the player that is assigned to play those URLs. So maybe it is a different one than the vlc player.

**Comment by Philipp91** [comment]
Yep, I had Google Photos set as the default. Some experiments showed that (1) it doesn't play my H264 MPEG-4 AVC files, MPEG-1/2 files are distorted and (2) it attempts to download the entire .ts file before starting playback, which would only work for small files while bigger files just resulted in a timeout.

Removing the "default" setting from Google Photos now lets me choose, and VLC player works fine. Thank you!

**Comment by rsiebert** [comment]
Just as an information: I use mxplayer (and vlc in the past) and when I call the url http://user:pass@tvheadend:9981/dvrfile/1843946285?ticket=7678BD7EF3FBD5260B780E3F048E98CFF9C5DCA0&mux=matroska (which is similar to yours above) the recording plays fine. 

You are the first one to report this issue. Does the user that is used in the tvhclient app has the permission to play / stream recordings? See http://docs.tvheadend.org/webui/config_access/

**Comment by jabbors** [comment]
Tried opening a recording in mxplayer but still wont play.

Could the ticket URL be a tvheadend setting? What version of tvheadend are you using? I'm using version 4.0.9 on Ubuntu 16.04.

**Comment by rsiebert** [comment]
The ticket stuff is a feature of tvheadend. I used to play recordings since the beginning of the app. At the beginning I was using tvheadend 3.4, 4.0. Now I use 4.1-2346. So its uptodate. Did you check the permissions?

**Comment by jabbors** [comment]
I checked the tvheadend logs and got the following

`2016-12-03 20:51:39.072 http: *.*.*.*: HTTP/1.1 GET /dvrfile/365711074?ticket=49BD98897DB9BB60CA254260B2CA893DC97CBD4D&mux=matroska -- 404`
`2016-12-03 20:51:39.177 http: *.*.*.*: HTTP/1.1 GET /favicon.ico -- 404`
`2016-12-03 21:03:13.186 http: *.*.*.*: using ticket 67E377A471EA44BD5E53F1B49829767B1F01A45B for /tvh/stream/channelid/13520387`

The first two entries are from trying to open a recording. No wonder it wont play when a 404 is returned.
The last entry is from opening a live channel stream. And it works fine.

I've setup tvheadend behind an HTTPS proxy so I start it up with `TVH_HTTP_ROOT="/tvh"` and as you can see the URL for the live channel stream is prefixed with `/tvh` whereas the recording URL is not. I bet this is the issue and thus the real issue is in the tvheadend service.

When I'm testing with the android client I'm connecting directly to tvheadend service on ports 9981-9982 as it doesn't support HTTPS yet. But the tvheadend service is still the same with a different HTTP root.

**Comment by jabbors** [comment]
If I start the service with `TVH_HTTP_ROOT=""` I can play both recordings and live streams.

**Comment by jabbors** [comment]
@rsiebert, can you confirm that you just add scheme, auth, host and port to the path obtained from a [getTicket](https://tvheadend.org/projects/tvheadend/wiki/Htsp#getTicket-Added-in-version-5) request when you build the URLs for live streams and recordings?

**Comment by rsiebert** [comment]
Yes. The url is http:user:pass@server:port/ ticket path ?ticket = ticket id

See https://github.com/rsiebert/TVHClient/blob/develop/app/src/main/java/org/tvheadend/tvhclient/PlayActivity.java#L95, https://github.com/rsiebert/TVHClient/blob/develop/app/src/main/java/org/tvheadend/tvhclient/PlayActivity.java#L353 and https://github.com/rsiebert/TVHClient/blob/develop/app/src/main/java/org/tvheadend/tvhclient/PlayActivity.java#L186

**Comment by jabbors** [comment]
@rsiebert 

As I assumed the webroot was missing in the path obtained from the ticket, so I hard coded tvheadend to verify.

First I tried this fix: https://github.com/jabbors/tvheadend/commit/25315ac2cd85cd1d6470b4c004bb16154161a52b, which made the recordings playable but not channel streams. In the log I found this `/tvh/tvh/stream/....`. So an extra webroot was added.

Then I removed it https://github.com/jabbors/tvheadend/commit/52091251467033a9d2ee1a1f201dd5457c93c8e4, which made both channel streams and recordings playable.

Do you add the webroot to channel stream URL?. I did not find any indication of it in the your code, but it would explain why I ended up with `/tvh/tvh/` instead of `/tvh/` in my first fix.

Also, according to this https://github.com/tvheadend/tvheadend/blob/04ff649a4c13229ebcaa9828fa04339a49842dd5/src/htsp_server.c#L1277, the hello message should contain the webroot, but I don't know if it's up to the client to include the webroot in the final URL or if the service shall provide the correct base path.

**Comment by jabbors** [comment]
Updated issue text. Link from the iOS client is NOT `https://*************@*************:9981/dvrfile/175c6fe5f3cb4d552c9670d0b3616a34`

**Comment by rsiebert** [comment]
In the getTicket method in https://github.com/rsiebert/TVHClient/blob/develop/app/src/main/java/org/tvheadend/tvhclient/htsp/HTSService.java#L1471 the webroot will be added to path. It is taken from the initial hello message. In the next bugfix release I'll add more logging. Then you can send me a debug log per mail.

**Comment by jabbors** [comment]
Great. Looking forward for the next release.

**Comment by rsiebert** [comment]
The new release 1.9.7 is out. It logs the webroot. You can now enable the debug log in the advanced settings and send me the mail.

**Comment by jabbors** [comment]
Sent you two log files.

The url for casting URL is still not correct. If the webroot is included in the URL is plays fine in VLC.

    Not working
    http://<user>:<pass>@<host>:9981/dvrfile/349928696?profile= 
    
    Working
    http://<user>:<pass>@<host>:9981/tvh/dvrfile/349928696?profile=

**Comment by rsiebert** [comment]
Added webroot in https://github.com/rsiebert/TVHClient/commit/ea69fa8b2dc508909bf3e14798830212c1074f37

**Comment by jabbors** [comment]
@rsiebert What's the reason for not prefixing the webroot in the casting url for recordings in your commit [ea69fa8](https://github.com/rsiebert/TVHClient/commit/ea69fa8b2dc508909bf3e14798830212c1074f37)?

**Comment by Osberg** [comment]
Nice, hope this fixes the issue with Chromecast also... :) when will a new release be out so we can test? :)

**Comment by rsiebert** [comment]
@jabbors Thx for pointing this out. I missed that. I'll fix this.

**Comment by rsiebert** [comment]
Added the missing webroot in https://github.com/rsiebert/TVHClient/commit/0d16ba24593fca42aa9068177c2145ff902c9e21

**Comment by Kledal** [comment]
@rsiebert Can you say anything about when it will be available on playstore? 😄

**Comment by rsiebert** [comment]
I think I will release a beta version this week. There are a couple of
other changes that I have added.

Alexander Kledal <notifications@github.com> schrieb am Di., 10. Jan. 2017,
22:27:

**Comment by Kledal** [comment]
I've just cloned the repo from the `developer` branch and is running in Android Studie. I still have the problem. I'm going to investigate the problem a bit further and maybe land a PR later

**Comment by rsiebert** [comment]
Any news here? The webroot was added in version 1.9.8.

**Comment by jabbors** [comment]
Still not working.

With the unpatched tvheadend binary only live channels was working. With a patched webroot tvheadend binary both live channels and recordings was working, but casting was not working.

I'll try to grab some log later.

**Comment by jabbors** [comment]
Sorry for not getting back earlier but we can mark this issue as resolved.

I upgraded my tvheadend server to version 4.2.x and now recordings plays just fine. I do however have a problem trying to cast recordings but I'll open a new issue for that.
